.github/workflows/deploy.yaml

```diff
@@ -29,45 +31,159 @@ jobs:
         with:
           path: ${{ github.event.repository.name }}
 
-      - name: Overwrite the global Rprofile ✍️
+      - name: Install jq
+        run: sudo apt-get update && sudo apt-get install -y jq
+
+      - name: Extract R version from lockfile
+        run: printf 'R_VERSION=%s\n' "$(jq --raw-output .R.Version ${{ github.event.repository.name }}/${{ matrix.directory }}/renv.lock)" >> $GITHUB_ENV
+
+      - name: Check if cypress test exists
+        id: find-cypress
+        run: |
+          if [ -d ${{ github.event.repository.name }}/${{ matrix.directory }}/tests/cypress ]; then
+            echo "has-cypress-tests=true" >> $GITHUB_OUTPUT
+          else
+            echo "Skipping frontend tests because cypress directory does not exist."
+            echo "has-cypress-tests=false" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Setup system dependencies for cypress and python app
+        if: steps.find-cypress.outputs.has-cypress-tests == 'true'
+        run: >
+          apt-get update && apt-get install --yes
+          libgtk2.0-0 libgbm-dev libnotify-dev libgconf-2-4 xvfb python3.10-venv
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        if: steps.find-cypress.outputs.has-cypress-tests == 'true'
+        with:
+          node-version: 16
+
+      - name: Setup npm packages
+        if: steps.find-cypress.outputs.has-cypress-tests == 'true'
+        shell: bash
+        working-directory: ${{ github.event.repository.name }}/${{ matrix.directory }}/js
+        run: |
+          npm i
+
+      - name: Restore renv from cache
+        uses: actions/cache@v2
+        env:
+          CACHE_KEY: renv-cache-${{ runner.arch }}-${{ runner.os }}-${{ env.R_VERSION }}-${{ matrix.directory }}-${{ env.CHANNEL }}
+        with:
+          path: ${{ github.event.repository.name }}/${{ matrix.directory }}/renv/library
+          key: ${{ env.CACHE_KEY }}-${{ hashFiles(format('{0}/{1}/renv.lock', github.event.repository.name, matrix.directory)) }}
+          restore-keys: ${{ env.CACHE_KEY }}-
+
+      - name: Update renv lock file with repo channel
+        shell: Rscript {0}
+        working-directory: ${{ github.event.repository.name }}
+        run: |
+          setwd("${{ matrix.directory }}")
+          repos <- renv::lockfile_read()$R$Repositories
+          repos$Pharmaverse <- "${{ matrix.repository }}"
+          renv::lockfile_modify(repos = repos) |> renv::lockfile_write()
+
+      - name: Install R packages using renv and update the renv snapshot
+        shell: Rscript {0}
+        working-directory: ${{ github.event.repository.name }}/${{ matrix.directory }}
         run: |
-          cat > /usr/local/lib/R/etc/Rprofile.site <<RPROFILE
-           options(
+          options(renv.config.cache.symlinks = FALSE)
+          lockfile_packages <- renv::lockfile_read()$Packages |> names()
+          renv::update(lockfile_packages)
+          renv::snapshot(packages = lockfile_packages, update = TRUE)
```

Comment:
Yeah, that's right. The better logic would be to update dependencies just from a particular repo, I'll check if something like this is possible.

Reply:
The new logic is to get the available packages from r-universe `available.packages(repos = "https://pharmaverse.r-universe.dev/")[,1]` and excluding them in the renv update and snapshots.

Reply:
As a result we have empty .lock files which makes existance of renv in question. All in all I am not super convinced to use r-universe at all. It doesn't store historical state of code, has some lags by definition and requires quite odd workarounds. I really would like to have it very simple and very easy to understand and maintain.

How about the following:
- For packages we want to do the updates (e.g. `teal` but not `dplyr`): let's install the latest release from GH and refer that in the .lock file
```
(...)
    "teal": {
      "Package": "teal",
      "Version": "0.14.0",
      "Source": "GitHub",
      "RemoteType": "github",
      "RemoteHost": "api.github.com",
      "RemoteUsername": "insightsengineering",
      "RemoteRepo": "teal",
      "RemoteRef": "v0.14.0",
      "RemoteSha": "bf6262bf7f3383966b5148aac5059929a0f2bdfd"
    },
```
- during CI we will do the updates as follows:
  - for channel devel - use HEAD
```
x <- renv::lockfile_read()
for (i in x$Packages) {
    if (i$Source == "GitHub") {
        renv::record(sprintf("%s/%s", i$RemoteUsername, i$Package))
    }
}
```
  - for channel stable - use remotes syntax
```
x <- renv::lockfile_read()
for (i in x$Packages) {
    if (i$Source == "GitHub") {
        renv::record(sprintf("%s/%s@*release", i$RemoteUsername, i$Package))
    }
}
```
Unfortunately, `renv::update()` is not working as I initially thought. It updates locally installed packages and does not touch .lock file at all. So my idea was to read it and update by hand all the packages that are coming from GH (see point above)
- autocommit only in from within stable channel

As a result, both user and CI will just call `restore()` when trying to open an app.

Reply:
The main problem I see with GitHub is that sometimes it might require the users to set PAT, I will test this logic without PAT and see if this requires no further steps for the users.

Fundamentally we do not really require `renv.lock` files and I agree that having a partially filled renv.lock file is not something easy to convey to end users. All we need is logic of installing packages in an isolated place. What if we just do a `renv::init()` and run the app from scratch? I see some challenges like will that require the users to restart the session but it should be possible to have some workaround.

Reply:
Yes, that is a valid statement. CI as well as user would need to set PAT to be able to pull the packages from GH. Here, GH is used to somehow "mark" which packages to update during devel deploy as well as indicate a (super straightforward) way how to do that. My goal was to avoid some additional configuration file (e.g. DESCRIPTION file; something similar to stage-deps config) but maybe that is unavoidable. Let's see...

> Fundamentally we do not really require renv.lock files and I agree that having a partially filled renv.lock file is not something easy to convey to end users. All we need is logic of installing packages in an isolated place. What if we just do a renv::init() and run the app from scratch? I see some challenges like will that require the users to restart the session but it should be possible to have some workaround.

Please note that this will work only if you have all the used packages pre-installed or available on the repos you already got. While this could be true for CI, I really doubt it's going to be true for a typical end-user from the outside who just wanted to try out stuff. One big requirement for a teal.gallery repo is to allow a convenient use by someone who is new to `teal` - that is: the current `restore_and_run()`.